Extend signetchallenge to set target block spacing

starius opened the pull request:
Inspired by https://github.com/bitcoin/bitcoin/pull/27446, this PR implements the proposal detailed in the comment https://github.com/bitcoin/bitcoin/pull/27446#issuecomment-1516600820.

## Rationale

Introduce the ability to configure a custom target time between blocks in a custom Bitcoin signet network. This enhancement enables users to create a signet that is more conducive to testing. The change enhances the flexibility of signet, rendering it more versatile for various testing scenarios. For instance, I am currently working on setting up a signet with a 30-second block time. However, this caused numerous difficulty adjustments, resulting in an inconsistent network state. Regtest isn't a viable alternative for me in this context since we prefer defaults to utilize our custom signet when configured, without impeding local regtest development.

## Implementation

If the challenge format is `OP_RETURN PUSHDATA<params> PUSHDATA<actual challenge>`, the actual challenge from the second data push is used as the signet challenge, and the parameters from the first push are used to configure the network. Otherwise the challenge is used as is.

Under the previous rules, such a signet challenge would always evaluate to `false`, suggesting that it is likely not in use by anyone. Updating bitcoind to a version that includes this change will not cause any disruptions — existing signet challenges will retain their original meaning without alteration.

The only parameter currently available is "target_spacing" (default 600 seconds). To set it, place `0x01<target_spacing as uint64_t, little endian>` in the params. Empty params are also valid. If other network parameters are added in the future, they should use `0x02<option 2 value>`, `0x03<option 3 value>`, etc., following the protobuf style.

Two public functions were added to `chainparams.h`:
  - `ParseWrappedSignetChallenge`: Extracts signet params and signet challenge from a wrapped signet challenge.
  - `ParseSignetParams`: Parses `<params>` bytes of the first data push.

Function `ReadSigNetArgs` calls `ParseWrappedSignetChallenge` and `ParseSignetParams` to implement the new meaning of signetchallenge.

The description of the flag `-signetchallenge` was updated to reflect the changes.

A new unit tests file, `chainparams_tests.cpp`, was added, containing tests for `ParseWrappedSignetChallenge` and `ParseSignetParams`.

The test `signet_parse_tests` from the file `validation_tests.cpp` was modified to ensure proper understanding of the new logic.

In the functional test `feature_signet.py`, a test case was added with the value of `-signetchallenge` set to the wrapped challenge, setting spacing to 30 seconds and having the actual challenge `OP_TRUE`.

The Signet miner was updated, introducing a new option `--target-spacing` with a default of 600 seconds. It must be set to the value used by the network.

## Example

I tested this PR against [Mutinynet](https://blog.mutinywallet.com/mutinynet/), a signet running on a custom fork of Bitcoin Core, implementing 30s target spacing. I successfully synchronized the blockchain using the following config:

```
signet=1
[signet]
signetchallenge=6a4c09011e000000000000004c25512102f7561d208dd9ae99bf497273e16f389bdbd6c4742ddb8e6b216e64fa2928ad8f51ae
addnode=45.79.52.207:38333
dnsseed=0
```

The content of this wrapped challenge:

```
6a OP_RETURN
4c OP_PUSHDATA1
09 (length of signet params = 9)
011e00000000000000 (signet params: 0x01, pow_target_spacing=30)
4c OP_PUSHDATA1
25 (length of challenge = 37)
512102f7561d208dd9ae99bf497273e16f389bdbd6c4742ddb8e6b216e64fa2928ad8f51ae
(original Mutinynet challenge)
```

starius committed:
Extend signetchallenge to set target block spacing

Inspired by https://github.com/bitcoin/bitcoin/pull/27446, this commit
implements the proposal detailed in the comment
https://github.com/bitcoin/bitcoin/pull/27446#issuecomment-1516600820.

Rationale.

Introduce the ability to configure a custom target time between blocks in a
custom Bitcoin signet network. This enhancement enables users to create a signet
that is more conducive to testing. The change enhances the flexibility of signet,
rendering it more versatile for various testing scenarios. For instance, I am
currently working on setting up a signet with a 30-second block time. However,
this caused numerous difficulty adjustments, resulting in an inconsistent
network state. Regtest isn't a viable alternative for me in this context since
we prefer defaults to utilize our custom signet when configured, without
impeding local regtest development.

Implementation.

If the challenge format is "OP_RETURN PUSHDATA<params> PUSHDATA<actual
challenge>", the actual challenge from the second data push is used as the
signet challenge, and the parameters from the first push are used to configure
the network. Otherwise the challenge is used as is.

Under the previous rules, such a signet challenge would always evaluate to
false, suggesting that it is likely not in use by anyone. Updating bitcoind to a
version that includes this change will not cause any disruptions - existing
signet challenges will retain their original meaning without alteration.

The only parameter currently available is "target_spacing" (default 600
seconds). To set it, place "0x01<target_spacing as uint64_t, little endian>" in
the params. Empty params are also valid. If other network parameters are added
in the future, they should use "0x02<option 2 value>", "0x03<option 3 value>",
etc., following the protobuf style.

Two public functions were added to chainparams.h:
  - ParseWrappedSignetChallenge: Extracts signet params and signet challenge
    from a wrapped signet challenge.
  - ParseSignetParams: Parses <params> bytes of the first data push.

Function ReadSigNetArgs calls ParseWrappedSignetChallenge and ParseSignetParams
to implement the new meaning of signetchallenge.

The description of the flag -signetchallenge was updated to reflect the changes.

A new unit tests file, chainparams_tests.cpp, was added, containing tests for
ParseWrappedSignetChallenge and ParseSignetParams.

The test signet_parse_tests from the file validation_tests.cpp was modified to
ensure proper understanding of the new logic.

In the functional test feature_signet.py, a test case was added with the value
of -signetchallenge set to the wrapped challenge, setting spacing to 30 seconds
and having the actual challenge OP_TRUE.

The Signet miner was updated, introducing a new option --target-spacing with a
default of 600 seconds. It must be set to the value used by the network.

Example.

I tested this commit against Mutinynet, a signet running on a custom fork of
Bitcoin Core, implementing 30s target spacing. I successfully synchronized the
blockchain using the following config:

signet=1
[signet]
signetchallenge=6a4c09011e000000000000004c25512102f7561d208dd9ae99bf497273e16f389bdbd6c4742ddb8e6b216e64fa2928ad8f51ae
addnode=45.79.52.207:38333
dnsseed=0

The content of this wrapped challenge:

6a OP_RETURN
4c OP_PUSHDATA1
09 (length of signet params = 9)
011e00000000000000 (signet params: 0x01, pow_target_spacing=30)
4c OP_PUSHDATA1
25 (length of challenge = 37)
512102f7561d208dd9ae99bf497273e16f389bdbd6c4742ddb8e6b216e64fa2928ad8f51ae
(original Mutinynet challenge, can be found here:
https://blog.mutinywallet.com/mutinynet/ )

benthecarman commented:
Huge Concept ACK, this is great

Sjors commented on src/kernel/chainparams.cpp:
```diff
@@ -341,7 +341,7 @@ class SigNetParams : public CChainParams {
         consensus.SegwitHeight = 1;
         consensus.nPowTargetTimespan = 14 * 24 * 60 * 60; // two weeks
-        consensus.nPowTargetSpacing = 10 * 60;
+        consensus.nPowTargetSpacing = options.pow_target_spacing;
```
My initial worry when seeing this PR was that it would touch mainnet consensus code in order to support these parameters. But clearly it doesn't, which is nice.

Eunovo commented on src/chainparams.cpp:
```diff
@@ -21,6 +22,72 @@
+        return;
+    }
+
+    if (params.size() != 1 + 8) {
```
Any specific reason why the `params` length must be 9?

starius replied:
The format of params is extendable in case more fields are added in the future. It is encoded as a concatenation of (field_id, value) tuples, protobuf style. Currently there is only one field defined `pow_target_spacing`, whose field_id is 0x01 and the lendth of encoding is 8 (int64_t). So valid values of params are:

  * empty string (checked in `if` block above)
  * 0x01 followed by 8 bytes of pow_target_spacing (9 bytes in total)

If length is not 0 and not 9, the value can not be parsed.

Eunovo replied:
@starius Might be useful to leave comment explaining why the params size is limited this way

TheCharlatan commented:
Concept ACK

Eunovo commented:
Tested ACK
Tried different `signetchallenge` inputs and got the expected results.

starius commented:
Rebased to resolve conflict.

edilmedeiros commented on test/functional/feature_signet.py:
```diff
@@ -28,7 +28,7 @@ def set_test_params(self):
         self.num_nodes = 6
         self.setup_clean_chain = True
-        shared_args1 = ["-signetchallenge=51"]  # OP_TRUE
+        shared_args1 = ["-signetchallenge=6a4c09011e000000000000004c0151"]  # OP_TRUE, target_spacing=30.
```
I don't feel like changing this test is a good idea since the proposed change is supposed to be backward-compatible. Better would be to add an additional functional test just for the purpose of testing the new behavior while keeping the old functional test as a guard rail.

starius replied:
Thanks for the proposal! I reworked the change of feature_signet.py test. Instead of modifying the existing test case, I added a new one with the wrapped signet challenge.

edilmedeiros reviewed (commented):
Concept ACK.

I prefer the approach in #27446 but I can't say this would not be the bitcoin style of using the existing script machinery to implement so a logic.

I would love to see an accompanying BIP to upgrade BIP 325, since the signet behavior is standardized there. In practice, this change in core will potentially make this implementation an "undocumented standard". 

I personally would prefer the BIP discussion to happen before merging, specially to define which params would be included in the configurable set and see if anyone proposes a better encoding. I'm not suggesting, though, that this doesn't get merged if consensus form around this proposal. I was messing around recently to set a faster signet for instructional purposes, this seem to be the perfect fit.

I'm still to give the code a better look and to test it locally.

carnhofdaki commented:
Concept ACK

cryptoquick commented:
Tested ACK
Works great against the Mutinynet node with the provided configuration, and the approach seems appropriate

starius commented:
I rebased the PR against latest `master`. Manually resolved a small code conflict in src/chainparams.cpp and a large one in contrib/signet/miner. In new OOP version of contrib/signet/miner my change is much smaller :)

Also I noticed that Makefile based testing was replaced with CMake based approach. I added new C++ file with tests (chainparams_tests.cpp) to src/test/CMakeLists.txt instead of src/Makefile.test.include.

torkelrogstad commented:
Concept ACK

starius commented:
@Zk2u Thanks for the patch! I applied it.

i-am-yuvi commented:
Great work!! @starius 
Concept ACK 

This would be a great help for Warnet Game since now you can easily configure the block creation time. Will review the code very soon!!

pinheadmz commented:
> This would be a great help for Warnet Game since now you can easily configure the block creation time. Will review the code very soon!!

Indeed, this is how we do it now:
https://github.com/bitcoin-dev-project/battle-of-galen-erso/blob/15ac31e0ec78dae716afc150f62e5cbc0fb7bd2e/scenarios/stub_orphan.py#L25-L31

i-am-yuvi commented:
> > This would be a great help for Warnet Game since now you can easily configure the block creation time. Will review the code very soon!!
> 
> Indeed, this is how we do it now: https://github.com/bitcoin-dev-project/battle-of-galen-erso/blob/15ac31e0ec78dae716afc150f62e5cbc0fb7bd2e/scenarios/stub_orphan.py#L25-L31

Interesting!! Thanks for sharing 🔥

i-am-yuvi commented:
https://github.com/bitcoin/bitcoin/issues/31446

This should fix the issue with CI failure!!

yuvicc reviewed (approved):
Tested ACK 6ce1b0ed7b8e4115f5ffed4c51c1d4eb0d6d5ddf

Tested using various signet challenges(30s, 1min, 5 mins, 30mins, etc), worked as expected.

starius commented:
I'll rebase after https://github.com/bitcoin/bitcoin/pull/31468 is merged to make CI green.

starius commented:
Rebased the PR, including https://github.com/bitcoin/bitcoin/pull/31468 and resolving the conflict in `test/functional/feature_signet.py`.

pinheadmz commented on test/functional/feature_signet.py:
```diff
@@ -80,20 +88,22 @@ def check_getmininginfo(node_idx, signet_idx):
-            assert_equal(self.nodes[2].submitblock(block), None)
-            height += 1
-            assert_equal(self.nodes[2].getblockcount(), height)
+        for node_idx in [2, 6]:
```
Why change this? Why not also include nodes 0 & 1?

starius replied:
I verify that nodes accept blocks mined using the default signet challenge. This test includes one default signet node (2) and another node (6) configured with an extended signet challenge that uses the actual script OP_TRUE (accepts all blocks).

I added a comment with this explanation to the code.

pinheadmz reviewed (commented):
concept ACK

pinheadmz commented:
> This change is backward-compatible, as per the old rules, such a signet challenge would always evaluate to false, rendering it unused.

I don't completely agree with this, since starting an old version of bitcoind with (for example) 
`-signetchallenge=6a4c0901ff000000000000004c0151` would create an incompatible node. It would have a different network magic bytes and I think the `OP_RETURN` challenge would cause it to reject all blocks.

starius commented:
> > This change is backward-compatible, as per the old rules, such a signet challenge would always evaluate to false, rendering it unused.
> 
> I don't completely agree with this, since starting an old version of bitcoind with (for example) `-signetchallenge=6a4c0901ff000000000000004c0151` would create an incompatible node. It would have a different network magic bytes and I think the `OP_RETURN` challenge would cause it to reject all blocks.

The change is backward-compatible in the sense that if bitcoind is updated to a version with this change, nothing will break - no existing signet challenge should have its meaning altered.

You are correct that starting an old node with an extended signet challenge will result in it rejecting all blocks. However, I believe this outcome is actually beneficial: it would alert the node owner to the issue, prompting them to update their software.

I updated that part of the description to make it more clear: "Under the previous rules, such a signet challenge would always evaluate to `false`, suggesting that it is likely not in use by anyone. Updating bitcoind to a version that includes this change will not cause any disruptions — existing signet challenges will retain their original meaning without alteration.".

> It would have a different network magic bytes.

This isn't accurate. The network magic bytes are derived solely from the real challenge (`51` (OP_TRUE) in the case of `6a4c0901ff000000000000004c0151`). Additionally, `getblockchaininfo` and `getmininginfo` both return `51`, not the extended challenge, as verified in the `feature_signet.py` test.

pinheadmz commented:
> This isn't accurate.

I meant, if you passed the extended challenge to an old version of bitcoin it would compute different magic bytes than if you passed the same challenge to an upgraded node.

1440000bytes reviewed (commented):
utACK https://github.com/bitcoin/bitcoin/pull/29365/commits/41728b516ec28ba2b6bbffd0b5e4cec8120f61d1

Isn't the commit message too long? You could even have multiple commits in this pull request IMO.

Sjors commented:
That sounds reasonable at first glance. Each commit has to pass the tests. Ideally anytime you introduce a new function you would also add at least one test, and/or use that function in existing code (to show that it doesn't break any test).

pinheadmz reviewed (commented):
code reviewed at 41728b516e

Built and tested on arm64/macos. Played with the feature locally using 10 and 30 second block times. A few comments below, my main concern being a mismatch between the python miner and the custom consensus rules. Didn't have time to try and mine 2, 016 signet blocks to check that difficulty adjusted as expected, but otherwise everything seems to work as expected.

Consensus code is touched, but only for signet ;-)

pinheadmz commented:
Crashed with long time spacing (are you sure you need 2^64 seconds?)

`build/bin/bitcoind -signet -signetchallenge=6a4c09011effff00000000004c0151 -debug -debugexclude=libevent`

`bitcoin-cli -signet -generate`

log:
```

2025-05-02T19:18:23Z [http] Received a POST request for / from 127.0.0.1:39282
2025-05-02T19:18:23Z [rpc] ThreadRPCServer method=getnewaddress user=__cookie__
2025-05-02T19:18:24Z [http] Received a POST request for / from 127.0.0.1:39284
2025-05-02T19:18:24Z [rpc] ThreadRPCServer method=generatetoaddress user=__cookie__
2025-05-02T19:18:24Z CreateNewBlock(): block weight: 884 txs: 0 fees: 0 sigops 400
Floating point exception
```

starius commented:
> Crashed with long time spacing (are you sure you need 2^64 seconds?)
> 
> `build/bin/bitcoind -signet -signetchallenge=6a4c09011effff00000000004c0151 -debug -debugexclude=libevent`
> 
> `bitcoin-cli -signet -generate`
> 
> log:
> 
> ```
> 
> 2025-05-02T19:18:23Z [http] Received a POST request for / from 127.0.0.1:39282
> 2025-05-02T19:18:23Z [rpc] ThreadRPCServer method=getnewaddress user=__cookie__
> 2025-05-02T19:18:24Z [http] Received a POST request for / from 127.0.0.1:39284
> 2025-05-02T19:18:24Z [rpc] ThreadRPCServer method=generatetoaddress user=__cookie__
> 2025-05-02T19:18:24Z CreateNewBlock(): block weight: 884 txs: 0 fees: 0 sigops 400
> Floating point exception
> ```

Great catch!

I think this is related to this place in `src/kernel/chainparams.cpp`:
```
consensus.nPowTargetTimespan = 14 * 24 * 60 * 60; // two weeks
consensus.nPowTargetSpacing = options.pow_target_spacing;
```

There is a code which divides these numbers in `src/consensus/params.h`:
```
int64_t DifficultyAdjustmentInterval() const { return nPowTargetTimespan / nPowTargetSpacing; }
```

`nPowTargetTimespan` should be calculated as `2016 * options.pow_target_spacing` so difficulty adjustment interval scales together with the block interval.  I'll apply the patch.

I'll try to reproduce this crash in `test/functional/feature_signet.py` adding edge cases.

> are you sure you need 2^64 seconds?

I don't think we need such high values :-)
Maybe 2 bytes is enough. It corresponds to ~18 hours.

starius commented:
I think I discovered a bug in existing test `test/functional/feature_signet.py`.
I found that it generates a block. I added the code to make sure that the block is distributed between the two nodes (node 0 and node 1) using the same signet challenge:
```diff
diff --git a/test/functional/feature_signet.py b/test/functional/feature_signet.py
index 02e37f0fdd..86ab3202b8 100755
--- a/test/functional/feature_signet.py
+++ b/test/functional/feature_signet.py
@@ -87,7 +87,10 @@ class SignetBasicTest(BitcoinTestFramework):
         check_getmininginfo(node_idx=3, signet_idx=1)
         check_getmininginfo(node_idx=4, signet_idx=2)
 
-        self.generate(self.nodes[0], 1, sync_fun=self.no_op)
+        while not self.generate(self.nodes[0], 1, sync_fun=lambda: self.sync_blocks(self.nodes[0:2])):
+            pass
+        self.wait_until(lambda: self.nodes[0].getblockcount() == 1)
+        self.wait_until(lambda: self.nodes[1].getblockcount() == 1)
 
         self.log.info("pregenerated signet blocks check")
```
(The patched branch can be found [here](https://github.com/starius/bitcoin/tree/mmm). It is the patch above applied to the `master` branch.)

The loop is needed because one `self.generate` call sometimes produces no blocks. My theory is that this happens because of `maxtries=1000000` passed by the test framework to the node. Sometimes 1000000 tries is not enough to mine a block. That is why I added a loop.

`self.wait_until` is needed to make sure both nodes got the block. But this code is not even executed if the test fails.

This made the test flaky. [Here is one of the fail logs](https://gist.githubusercontent.com/starius/7f090a314002d0eeb5371b05a814c116/raw/9b4286a92cb5dc5fa3d0d020df7eb2cf0f3cbec2/gistfile1.txt)

I see that node 0 successfully generated the block and updated its tip:
```
node0 2025-05-05T04:55:18.225539Z [scheduler] [validationinterface.cpp:179] [operator()] [validation] UpdatedBlockTip: new block hash=000002d68e882daae449c4eee999af64d893c68e34840737daa32d6d557dc5fb fork block hash=00000008819873e925422c1ff0f99f7cc9bbb232af63a077a480a3633bee1ef6 (in IBD=false) 
```

but node 1 for some reason got only block header and not the block itself:
```
node1 2025-05-05T04:55:18.226722Z [msghand] [net_processing.cpp:3437] [LogBlockHeader] [validation] Saw new header hash=000002d68e882daae449c4eee999af64d893c68e34840737daa32d6d557dc5fb height=1 peer=0 
```

Then `self.sync_blocks` failed because of a timeout.

How to debug this further? Is this a genuine bug or my change above is wrong?
Any help to sort this out is very appreciated!

pinheadmz commented:
What commit is that failure from?

JeremyRubin commented on src/chainparams.cpp:
```diff
@@ -21,8 +22,83 @@
+    //  - 0x01 followed by 8 bytes of pow_target_spacing (9 bytes in total).
+    // If length is not 0 and not 9, the value can not be parsed.
+
+    if (params.size() != 1 + 8) {
```
this should come _after_ the version check

JeremyRubin commented on src/chainparams.cpp:
```diff
@@ -21,8 +22,83 @@
+    // Parse little-endian 64 bit number to uint64_t.
+    const uint8_t* bytes = &params[1];
+    const uint64_t value = uint64_t(bytes[0]) | uint64_t(bytes[1])<<8 | uint64_t(bytes[2])<<16 | uint64_t(bytes[3])<<24 |
+            uint64_t(bytes[4])<<32 | uint64_t(bytes[5])<<40 | uint64_t(bytes[6])<<48 | uint64_t(bytes[7])<<56;
```
i would probably make parsing work by doing a SERIALIZE_METHODS on a struct, and have any extensions in the future be done by failing to deserialize unknown versions...

JeremyRubin commented:
Concept ACK.

We use something like this for testing and it's quite helpful.

it's nice being able to ship a parameter pack in a single hex for a signet.

I think the future version flexibility can be simplified, reducing @Sjors concern of the wrapper complexity. A good ol' serialize_methods chunk can probably make this trivial :)

e.g.:


```c++
    SERIALIZE_METHODS(WrappedSignetChallenge, obj)
    {
        READWRITE(obj.version);
        if (obj.version == 0) {
            READWRITE(obj./*...*/);
            // ....
        } else {
              if (ser_action.ForRead()) {
                   throw std::ios_base::failure(
                strprintf("Attempted to Read Unknown WrappedSignetChallenge", obj.version))
             }
             if (ser_action.ForWrite()) {
                 assert(0 && "Unknown Version Generated By This Node -- please add serialization code to support new versions");
             }
    }

```
